### PR TITLE
[REF] web: data.js: refactoring of BufferedDataSet

### PR DESCRIPTION
**Pull request by Gorash**
Add a test to check the one2many create form and onchange
Refactoring of BufferedDataSet (less code and speed up)

### CODE REVIEW

File: addons/web/static/src/js/framework/data.js [hunk 1]
```diff
@@ -739,155 +739,92 @@ var BufferedDataSet = DataSetStatic.extend({
             self.last_default_get = res;
         });
     },
+    get_cache: function (id) {
+        if (!this.cache[id]) {
+            this.cache[id] = {
+                'id': id, values: {},
```
Comment (odony) [inline addons/web/static/src/js/framework/data.js]: reindent, quote

File: addons/web/static/src/js/framework/data.js [hunk 2]
```diff
@@ -739,155 +739,92 @@ var BufferedDataSet = DataSetStatic.extend({
...
+                'from_read': {},
+                'change': {},
```
Comment (odony) [inline addons/web/static/src/js/framework/data.js]: -> `changes`

File: addons/web/static/src/js/framework/data.js [hunk 3]
```diff
@@ -997,9 +922,9 @@ var BufferedDataSet = DataSetStatic.extend({
         // Don't evict records which haven't yet been saved: there is no more
         // recent data on the server (and there potentially isn't any data),
         // and this breaks the assumptions of other methods (that the data
-        // for new and altered records is both in the cache and in the to_write
+        // for new and altered records is both in the cache and in the change
         // or to_create collection)
-        this.from_read = _.reject(this.from_read, function(x) { return x.id === id;});
+        this._update_cache(id).from_read = {};
```
Comment (odony) [inline addons/web/static/src/js/framework/data.js]: merge the 2 lines

File: addons/web/static/src/js/views/form_relational_widgets.js [hunk 4]
```diff
@@ -573,7 +570,7 @@ var AbstractManyField = common.AbstractField.extend({
         var def = $.Deferred();
         var dataset = this.dataset;
         var res = true;
-        options = options || {};
+        var _options = _.extend({'internal_dataset_changed': true}, options);
```
Comment (odony) [inline addons/web/static/src/js/views/form_relational_widgets.js]: `_options` -> `internal_options`

File: addons/web/static/src/js/views/form_relational_widgets.js [hunk 5]
```diff
@@ -682,8 +671,17 @@ var AbstractManyField = common.AbstractField.extend({
         return command_list;
     },
 
+    _format_values : function (values) {
```
Comment (odony) [inline addons/web/static/src/js/views/form_relational_widgets.js]: keep previous name (`_format_many2one`) and clone() inside (or check if an util method exists to do the same)

**Comment by Gorash** [comment]
k

**Pull request closed without merging**